xdg-xmenu.1
.TH XDG-XMENU 1 "Jun 2023" VERSION

.SH NAME
xdg-xmenu - show app menu using
.IR xmenu (1)

.SH SYNOPSIS
.B xdg-xmenu
.RB [ -dGIn ]
.RB [ -b
.IR fallback_icon ]
.RB [ -i
.IR icon_theme ]
.RB [ -s
.IR icon_size ]
.RB [ -S
.IR scale ]
.RB [ -t
.IR terminal ]
.RB [ -x
.IR xmenu_cmd ]
.RB [ --
.IR <xmenu_options> ]

.SH DESCRIPTION
.B xdg-xmenu
is a small program that shows an app menu using
.IR xmenu (1).
It tries to follow the specifications of desktop entries and icon themes
provided by freedesktop:
.IP
https://specifications.freedesktop.org/desktop-entry-spec
.IP
https://specifications.freedesktop.org/icon-theme-spec
.P
.B xdg-xmenu
also has the potential to generate menu for other general-purpose
menu utilities.

.SH OPTIONS
.TP
.BI -b " fallback_icon"
Fallback icon in case one can not be found.
Accept either an icon name or a file path.
Default is application-x-executable.
.TP
.B -d
Print the menu to stdout and exit, do not run
.IR xmenu (1)
command.
.TP
.B -G
Do not show app's generic names.
.TP
.BI -i " icon_theme"
Icon theme.
Default value is parsed from gtk3 configuration file.
Fallback to "hicolor" if none are found.
.TP
.B -I
Disable icons. This implies an '-i' option passed to
.IR xmenu (1)
, which also disables icons display
to make it loading faster.
.TP
.B -n
Dry run mode. Do not run the selected app. Instead, the selection will be
printed to stdout, as in the behavior of vanilla xmenu.
.TP
.BI -s " icon_size"
Icon size. This is used when searching for icon files. It's not xmenu's display
size. Default is 24.
.TP
.BI -S " scale"
Scale of icons. Useful in HiDPI screens. For example, 24x24@2x icons looks like
24x24 icons, but with sharper details. While 48x48 icons is at the same
resolution, but possibly with more details.
.TP
.BI -t " terminal"
Terminal emulator to use. Default is xterm.
.TP
.BI -x " xmenu_cmd"
Alternative xmenu command to use, default is xmenu. This is only the
executable, extra command should be passed after `--', see below.
.TP
.B --
Everything after `--' are passed to `xmenu_cmd' command.

.SH EXAMPLES
.P
The -x and -- two options provides infinite possiblity, although not so useful
for now. There is a redundant example equivlant to `xdg-xmenu -I':
.IP
xdg-xmenu -I -x xmenu -- -i
.P
This disables both icons in
.B xdg-xmenu
and
.IR xmenu (1).

.SH RESOURCES
.SS Desktop Files
The script will go through the .desktop files in the following directories:
.IP
$XDG_DATA_DIRS/applications
.IP
$XDG_DATA_HOME/applications
.P
This covers the cases like flatpak, where the flatpak-specific folders
will be appended to the XDG_DATA_DIRS environment variable (by flatpak).
So this program can find them, too.
.SS Icon Files
.B
xdg-xmenu
will find the icons in the following directories (loosely
following the method in the freedesktop.org specification):
.IP
$XDG_DATA_DIRS/icons
.IP
$XDG_DATA_HOME/icons

.SH HISTORY
.P
This program started out to be a shell script. But since the speed was quite
slow (~500ms), the user could only export the menu file, and feed it to xmenu.
.P
Then I wrote a python version of this program. Although python itself is fast
enough, but the startup time takes up majority of the execution time (~50ms).
.P
Finally, I decided to write a C version, even though I am not that fluent in C.
It shouldn't be hard to do, since it mostly involves text processing and file
searching. Finally, with about double the lines of code (SLOC), this version
works quite well and fast. The average time is about 10ms on my machine. The
bottleneck is now file access.

.SH BUGS
The implementation does not cover everything in the freedesktop specifications.
For example the complicated icon search algorithm, and recursive app folder.
So it is possible that a desktop entry is parsed incorrectly or an icon is not
found. However in my own experience, it is better than most menu utilities.
.P
Report any bugs at
.IR https://github.com/xlucn/xdg-xmenu/issues

.SH AUTHOR

Lu Xu <oliver_lew at outlook dot com>

.SH SEE ALSO
.IR xmenu (1)


CHANGELOG
# C version

This current version of xdg-xmenu is written in C, before which there was a
Python version and a shell version (see below for their history versions).
The options are different between different variations, so modify your command
after installing a new one.

This C version can run really fast (e.g., about 10 to 20 ms on my machine) so
that one can directly start xmenu with it. In contrast, a user certainly had to
export the menu to a file first and then feed the menu to xmenu later, if he is
using the shell version.


v1.0.0-beta.1 2022-10-28

Added:
- Initial C version.


======

# Python version

Below is the legacy Python version. Later C version will not be backward
compatible with this in terms of cli interface. Although I will try to
make most the options the same.

The python versions are just intermediate state of this program, which I never
introduced to anyone. I wrote the C version right after this. Please use that
instead.


v0.99.1 2022-09-20

Fixed:
- Order of the categories.


v0.99.0 2022-08-07

Added:
- Initial Python version.


======

# Shell version

Below is the legacy shell version. Later Python and C versions will not be
backward compatible with this in terms of cli interface. Although I will try to
make most the options the same.


0.4.2 2022-05-20

Added:
- Option `-G` to hide generic name. #8


0.4.1 2022-05-07

Changed:
- Hide empty categories by default, use option `-E` to show them. #7

Fixed:
- Check for empty applications folder. #6


0.4.0 2022-03-01

Added:
- Option `-S` to use svg icons directly with the new Imlib2 1.8.0.

Fixed:
- Threshold check logic


0.3.1 2021-10-01

Improved:
- Better icon size matching.

Fixed:
- Won't exit after printing the help message.
- Possible conflict of cached image file. #4


0.3.0 2021-09-16

Added:
- `-f` option to force caching icons even if already cached.
  Useful when switching to a new icon theme, or the icon theme get updated.
- `-n` option to skip icon caching, a.k.a. dry run mode.

Changes:
- Png and xpm icon images are not copied to the cache directory.
  This could slow down things a little to search for the icons everytime.
- Keep `.svg` for the generated image from svg->png convertion.
- Use `gtk-query-settings` command to query gtk settings instead from file.
  The fallback icon theme will be "Adwaita".


0.2.0 2021-05-06

Added:
- Add environment variables.

Fixed:
- Does not exit anymore if fallback icon can not be found. #1
- Fix empty IMG: tag. #2
- Fix file creation error when Icon key is a full path. #3


0.1.0 2020-08-05

Initial release.


xdg-xmenu.py
#!/usr/bin/env python
# Author: Lu Xu <oliver_lew at outlook dot com>
# License: MIT
# References: https://specifications.freedesktop.org/desktop-entry-spec
#             https://specifications.freedesktop.org/icon-theme-spec

import os
import argparse
import configparser


def get_args():
    parser = argparse.ArgumentParser(description="A python rewrite of the shell version")
    parser.add_argument('-b', '--fallback-icon', default="application-x-executable",
                        help='Fallback icon for apps without icons')
    parser.add_argument('-i', '--icon-theme', default="Adwaita",
                        help='Icon theme for app icons')
    parser.add_argument('-s', '--icon-size', type=int, default=24,
                        help='Icon theme for app icons')
    parser.add_argument('-e', '--xdg-de', action='store_true',
                        help='Show apps according to desktop environments')
    parser.add_argument('-N', '--dry-run', action='store_true',
                        help='Do not run app, output to stdout')
    parser.add_argument('-t', '--terminal', metavar='terminal', default='xterm',
                        help='Terminal emulator to use, default is %(default)s')
    return parser.parse_args()


def valid_dirs(theme_dir):
    # Find directories in an icon theme matching the target size
    indexfile = f"{theme_dir}/index.theme"
    if not os.path.exists(indexfile):
        return []

    # parse the index.theme file
    entries = configparser.RawConfigParser()
    entries.read(indexfile)
    for subdir in entries.get("Icon Theme", "Directories").rstrip(",").split(","):
        # "Size" is required
        Size = entries.getint(subdir, "Size")
        # FIXME: scaled icons are ignored for now, sorry HiDPI
        if entries.getint(subdir, "Scale", fallback=1) != 1:
            continue

        # defaults if they are not specified
        Type = entries.get(subdir, "Type", fallback="Threshold")
        MaxSize = entries.getint(subdir, "MaxSize", fallback=Size)
        MinSize = entries.getint(subdir, "MinSize", fallback=Size)
        Threshold = entries.getint(subdir, "Threshold", fallback=2)

        # match subdirectory sizes based on 'Type'
        if ((Type == "Threshold" and abs(args.icon_size - Size) <= Threshold) or
                (Type == "Scalable" and MinSize <= args.icon_size <= MaxSize) or
                (Type == "Fixed" and Size == args.icon_size)):
            yield f"{theme_dir}/{subdir}"


def find_icon_dirs():
    # Find all directories matching the size and containing icons directly
    all_dirs = []
    for datadir in DATA_DIRS:
        for theme in [args.icon_theme, "hicolor"]:
            all_dirs.extend(valid_dirs(f"{datadir}/icons/{theme}"))
    all_dirs.append("/usr/share/pixmaps")
    return all_dirs


def find_icon(icon, fallback):
    # Search for an icon based on its name
    # find any png, svg or xpm icon in available directories
    for dir in icon_dirs:
        for ext in ["png", "svg", "xpm"]:
            icon_file = f"{dir}/{icon}.{ext}"
            if os.path.exists(icon_file):
                return icon_file
    # when the fallback icon is not found, return none instead
    return None if icon == fallback else fallback


def gen_entry(entry):
    Exec = entry.get('Exec')
    Icon = entry.get('Icon')
    Name = entry.get('Name')
    Terminal = entry.get('Terminal')
    GenericName = entry.get('GenericName')
    Categories = entry.get('Categories')

    cmd = Exec.replace('%f', '') \
              .replace('%F', '') \
              .replace('%u', '') \
              .replace('%U', '') \
              .replace('%c', Name) \
              .replace('%k', entry.get('Path', ''))
    if Terminal == "true":
        cmd = f"{args.terminal} -e {cmd}"

    icon = find_icon(Icon, fallback)

    name = f"{Name} ({GenericName})" if GenericName else Name

    category = "Others"
    for c in Categories.rstrip(";").split(";"):
        if categories.get(c):
            category = c

    return {"category": category, "icon": icon, "name": name, "cmd": cmd}


def if_show(entry):
    if (entry.get('Exec') is None or
            entry.get('NoDisplay') == 'true' or
            entry.get('Hidden') == 'true' or
            entry.get('Type') != "Application"):
        return False

    if entry.get('TryExec'):
        TryExec = entry.get('TryExec')
        if os.path.isabs(TryExec):
            if os.path.exists(TryExec):
                return True
        else:
            for path in os.getenv("PATH").split(":"):
                if os.path.exists(os.path.join(path, TryExec)):
                    return True
        return False

    if not args.xdg_de:
        if entry.get('NotShowIn'):
            for DE in entry.get('NotShowIn').rstrip(';').split(';'):
                if current_desktops.find(DE) >= 0:
                    return False

        if entry.get('OnlyShowIn'):
            for DE in entry.get('OnlyShowIn').rstrip(';').split(';'):
                if current_desktops.find(DE) >= 0:
                    return True
            return False

    return True


def get_apps():
    apps = {}
    for data_dir in DATA_DIRS:
        app_dir = os.path.join(data_dir, 'applications')
        # Desktop entry files can be in nested directories
        for root, dirs, files in os.walk(app_dir):
            for name in files:
                if name.endswith('.desktop'):
                    app = os.path.join(root, name)
                    app_id = os.path.relpath(app, app_dir).replace('/', '-')

                    # Apps with the same id, choose the first
                    apps[app_id] = app
    return apps.values()


if __name__ == "__main__":
    args = get_args()
    current_desktops = os.getenv('XDG_CURRENT_DESKTOP', '')

    # XDG directories
    HOME = os.getenv("HOME")
    XDG_DATA_HOME = os.getenv("XDG_DATA_HOME", f"{HOME}/.local/share")
    XDG_DATA_DIRS = os.getenv("XDG_DATA_DIRS", "/usr/share:/usr/local/share").split(":")
    DATA_DIRS = [*XDG_DATA_DIRS, XDG_DATA_HOME]

    icon_dirs = find_icon_dirs()
    fallback = find_icon(args.fallback_icon, None)

    categories = {
        "AudioVideo": {"name": "Multimedia", "icon": "applications-multimedia"},
        "Audio": {"name": "Multimedia", "icon": "applications-multimedia"},
        "Video": {"name": "Multimedia", "icon": "applications-multimedia"},
        "Development": {"name": "Development", "icon": "applications-development"},
        "Education": {"name": "Education", "icon": "applications-education"},
        "Game": {"name": "Games", "icon": "applications-games"},
        "Graphics": {"name": "Graphics", "icon": "applications-graphics"},
        "Network": {"name": "Internet", "icon": "applications-internet"},
        "Office": {"name": "Office", "icon": "applications-office"},
        "Science": {"name": "Science", "icon": "applications-science"},
        "Settings": {"name": "Settings", "icon": "preferences-desktop"},
        "System": {"name": "System", "icon": "applications-system"},
        "Utility": {"name": "Accessories", "icon": "applications-accessories"},
        "Others": {"name": "Others", "icon": "applications-other"},
    }

    menu = {}
    for app in get_apps():
        app_parser = configparser.RawConfigParser()
        app_parser.read(app)
        entry = app_parser["Desktop Entry"]
        if if_show(entry):
            item = gen_entry(entry)
            if menu.get(item["category"]):
                menu[item["category"]].append(item)
            else:
                menu[item["category"]] = [item]

    for category in menu:
        category_name = categories[category]["name"]
        category_icon = categories[category]["icon"]
        icon_file = find_icon(category_icon, fallback)
        print(f"IMG:{icon_file}\t{category_name}")
        for item in menu[category]:
            print("\tIMG:{icon}\t{name}\t{cmd}".format_map(item))
